src/fbas_analyze.rs

```diff
@@ -0,0 +1,253 @@
+    }
+
+    pub(crate) fn from_fbas(fbas: Fbas, cb: Cb) -> Result<Self, FbasError> {
+        let mut analyzer = Self {
+            fbas,
+            solver: Solver::new(Default::default(), cb),
+            status: SolveStatus::UNKNOWN,
+        };
+        analyzer.construct_formula()?;
+        Ok(analyzer)
+    }
+
+    fn construct_formula(&mut self) -> Result<(), FbasError> {
+        let fbas = &self.fbas;
+        let fbas_lits = FbasLitsWrapper::new(fbas.graph.node_count());
+
+        // for each vertex in the graph, we add a variable representing it
+        // belonging to quorum A and quorum B
+        fbas.graph.node_indices().for_each(|_| {
+            self.solver.new_var_default();
+            self.solver.new_var_default();
+        });
+        // sanity check
+        if self.solver.num_vars() as usize != fbas.graph.node_count() * 2 {
+            return Err(FbasError::InternalError("variables and nodes do not match"));
+        }
+
+        // formula 1: both quorums are non-empty -- at least one validator must
+        // exist in each quorum
+        let mut quorums_not_empty: (Vec<Lit>, Vec<Lit>) = fbas
+            .validators
+            .iter()
+            .map(|ni| (fbas_lits.in_quorum_a(ni), fbas_lits.in_quorum_b(ni)))
+            .collect();
+        self.solver.add_clause_reuse(&mut quorums_not_empty.0);
+        self.solver.add_clause_reuse(&mut quorums_not_empty.1);
+
+        // formula 2: two quorums do not intersect -- no validator can appear in
+        // both quorums
+        fbas.validators.iter().for_each(|ni| {
+            self.solver.add_clause_reuse(&mut vec![
+                !fbas_lits.in_quorum_a(ni),
+                !fbas_lits.in_quorum_b(ni),
+            ]);
+        });
+
+        // formula 3: qset relation for each vertex must be satisfied
+        let mut add_clauses_for_quorum_relations =
+            |in_quorum: &dyn Fn(&NodeIndex) -> Lit| -> Result<(), FbasError> {
+                fbas.graph.node_indices().try_for_each(|ni| {
+                    let aq_i = in_quorum(&ni);
+                    let nd = fbas
+                        .graph
+                        .node_weight(ni)
+                        .ok_or_else(|| FbasError::InternalError("Node index not found"))?;
+                    let threshold = nd.get_threshold();
+                    let neighbors = fbas.graph.neighbors(ni);
+                    let qset = neighbors.into_iter().combinations(threshold as usize);
+
+                    let mut third_term = vec![];
+                    third_term.push(!aq_i);
+                    for (_j, q_slice) in qset.enumerate() {
+                        // create a new proposition as per Tseitin transformation
+                        let xi_j = fbas_lits.new_proposition(&mut self.solver);
+
+                        // this is the second part in the qsat_i^{A} equation
+                        let mut neg_pi_j = vec![];
+                        neg_pi_j.push(!aq_i);
```

Comment:
I think this is wrong. qa_i should only appear in the clause [!aq_i, xi_1, xi_2, ... xi_n]

Reply:
The formula has been updated. Resolving this comment.